### ISSUE TITLE
lightning should cache the result of `columnAPI.Cols()` to improve performance

### ISSUE THREAD

**Issue by lance6716** [pingcap/tidb#56705]
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

capture a CPU profile when lightning encodes KV

![image](https://github.com/user-attachments/assets/f488bf6d-e9dd-494b-a673-1b7ab64ec3d7)

https://github.com/pingcap/tidb/blob/ff2feb6653846b0decd8b1d7cc0a665e128ccf26/br/pkg/lightning/backend/kv/sql2kv.go#L491-L497

some implementations of `columnAPI.Cols()` just return the slice, like before https://github.com/pingcap/tidb/pull/50062 or after https://github.com/pingcap/tidb/pull/53798. The other will filter and clone the slice.

### 2. What did you expect to see? (Required)

stable performance for different versions of lightning

### 3. What did you see instead (Required)

the duration doubles

### 4. What is your TiDB version? (Required)

at least v6.5.10

**Comment by princejha95** [comment]
@lance6716 can i pick this up ?

**Comment by princejha95** [comment]
I am using tidb as backend and while trying to create a task, i am getting following error:

["encode kv data and write failed"] [table=test.posts] [engineNumber=0] [takeTime=316.715µs] [error="file 'test.posts-schema.sql' with unknown source type 'table-schema'"]

Do i need to change the source type ?

**Comment by lance6716** [comment]
please follow this doc https://docs.pingcap.com/tidb/stable/tidb-lightning-data-source to prepare the source data. Or you can take a look at lightning integration tests https://github.com/pingcap/tidb/tree/master/lightning/tests

**Comment by princejha95** [comment]
@lance6716 I am running the code from latest master branch and i am using **tidb** as backend. I see that the duration of encoding sql -> kv is around 2 mins.

![encode_logs](https://github.com/user-attachments/assets/fed34b74-7e51-4d42-9e2e-de1be184acb6)

Am i missing something here ?

Payload i am passing while creating task:

[lightning]
table-concurrency = 1
index-concurrency = 1
region-concurrency = 1
io-concurrency = 1
check-requirements = true
meta-schema-name = "meta-test"

[tidb]
host = "127.0.0.1"
port = 4000
user = "root"
status-port = 42483

**Comment by lance6716** [comment]
Maybe your disk or CPU resource is not enough. I see the `restore file completed` line `encodeDur` is too high for a `size=63` task.

**Comment by princejha95** [comment]
Yeah.. it takes ~29 secs to encode a task of size 63 bytes. Now that i have reproduced the issue, i will start working on the fix.

**Comment by lance6716** [comment]
> Yeah.. it takes ~29 secs which is unusually high to encode a task of size 63 bytes. Now that i have reproduced the issue, i will start working on the fix.

I think the time cost is not normal, maybe check your develop environment first? I'm a bit worried that it will affect your choice of fixing the issue. You can run a golang CPU profiling https://pkg.go.dev/runtime/pprof for that function to see why it's slow. See the example in that link, simply add `pprof.StartCPUProfile(f)` and `pprof.StopCPUProfile()` to cover the function that need to be checked.

**Comment by princejha95** [comment]
Sure. I will run the CPU profiling for the Encode function and verify what's the issue.

**Comment by lance6716** [comment]
@princejha95 Hi, we plan to fix it on our side if you have no further progress

**Comment by D3Hunter** [comment]
after https://github.com/pingcap/tidb/pull/50062 (since v7.6.0) which mentioned in the issue, `Cols` always clone. the PR is CP back to 6.5 branch, so it only affects [v6.5.8](https://github.com/pingcap/tidb/releases/tag/v6.5.8) - [v6.5.11](https://github.com/pingcap/tidb/releases/tag/v6.5.11)

since 7.1.0, in https://github.com/pingcap/tidb/pull/42689/files#diff-637b3e18d1551196f5d01d7b6026936b74ff19b9cdba05cf06926ac76cd5dbc2R161 the result of `Cols` is also cached, so no such issue too

so only lightning of version `[v6.5.8, v6.5.11]` have this issue

**Comment by OliverS929** [comment]
> after [#50062](https://github.com/pingcap/tidb/pull/50062) (since v7.6.0) which mentioned in the issue, `Cols` always clone. the PR is CP back to 6.5 branch, so it affects [v6.5.8](https://github.com/pingcap/tidb/releases/tag/v6.5.8) - [v6.5.11](https://github.com/pingcap/tidb/releases/tag/v6.5.11)
> 
> since 7.1.0, in https://github.com/pingcap/tidb/pull/42689/files#diff-637b3e18d1551196f5d01d7b6026936b74ff19b9cdba05cf06926ac76cd5dbc2R161 the result of `Cols` is also cached, so no such issue too
> 
> so only lightning of version `[v6.5.8, v6.5.11]` have this issue

Thank you for clarifying the issue.
